### ISSUE TITLE
Create direct download capability for podcast media files in article view

### ISSUE THREAD

**Issue by Eucyon** [newsboat/newsboat#562]
This would make it possible to bypass Podboat.

**Comment by Minoru** [comment]
Not convinced. What's wrong with Podboat, exactly?

**Comment by Eucyon** [comment]
Pressing one key to download a file is easier than pressing one key to add it to the queue, opening Podboat, pressing another key to downoad and pressing a key to delete the file from the queue.
 

Gesendet: Dienstag, 02. Juli 2019 um 20:37 Uhr
Von: "Alexander Batischev" <notifications@github.com>
An: newsboat/newsboat <newsboat@noreply.github.com>
Cc: Eucyon <martin_wolf@gmx.de>, Author <author@noreply.github.com>
Betreff: Re: [newsboat/newsboat] Create direct download capability for podcast media files in article view (#562)


Not convinced. What's wrong with Podboat, exactly?

**Comment by Minoru** [comment]
This sound like you're using Podboat just to download stuff. If that's the case, you might be better of writing a script that will watch the queue file, and on every change: parse it, download the files, and clear the queue. Perhaps this should be implemented in Podboat itself, even. Would you keep Podboat open just for it to automatically download podcasts?

I still don't think this is a good feature to have in Newsboat proper. With any podcast, user's end goal is to listen to it. This means not just downloading it, but also playing it; or perhaps notifying some other program that the file is ready. Then there are people (like me) who download podcasts in batches, and listen to the backlog as time permits; those people will need some way to manage their downloaded files.

**Comment by Eucyon** [comment]
Thanks for considering this again.

In my use case, I download quite a lot of audio files to my Download folder. So I am able to listen to them when I would like to.

I also thought about what you suggested ("writing a script..."). But in my eyes it would be easier to simply press a key after I have read the article in Newsboat and decided to download. This would be an additional capability of NB, which would make invoking Podboat unnecessary in my use case. Maybe this would be useful for others.

**Comment by Minoru** [comment]
With a script, you'll press exactly one key—'e'. Or even none at all, if auto-enqueue is enabled. But that script still needs to be written.

You didn't answer if you would run Podboat in background if it could automatically download stuff, obviating the need for the script. Would you?

I'm not trying to persuade you that you don't need this feature. Instead, I'm trying to show that this is a case of feature creep: adding just one key that downloads podcasts won't be enough, and eventually we'll have to move the whole of Podboat into Newsboat. The result will be harder to maintain, especially since Podboat doesn't seem to enjoy any popularity at all—can't remember getting any patches for it, unlike hundreds of PRs to Newsboat.

In short, I'd rather see Podboat improved to a point where it covers your use-case, rather than turn Newsboat into even bigger of a monster.

**Comment by Eucyon** [comment]
OK, I will think about a way, maybe via letting PB run in background.

Thanks!

**Comment by Minoru** [comment]
Wait, there is no point *today* in running Podboat in background. My question was about a hypothetical future where Podboat learns to do watch the queue and automatically download everything it sees. I'm trying to figure out how I can improve Podboat to suit your needs better, so let's think through this together please?

**Comment by Eucyon** [comment]
Thanks for accomodating.

But still, I would prefer not to be forced to invoke a second program to download a media file.

I saw that the queue is a plain file containing the URLs in plaintext. So I probably will write a little script that I start at the same time as NB that watches the queue, downloads the files in it and exits when NB is closed.

**Comment by Eucyon** [comment]
This script watches NB's queue in background, starts NB, downloads all files added to the queue to the "Download" folder and quits, when NB is quit:

```
#!/bin/bash

{
  sleep 5
  while pidof newsboat > /dev/null
  do
    while read url filename; do
      wget -qP ~/Downloads "$url" && sed -i "\!^$url!d" ~/.newsboat/queue
    done < ~/.newsboat/queue
    sleep 5
  done
} &

newsboat
```

It depends on wget.

Put it in the binary folder (~/bin) and start it with the name you gave it, e.g "n".

**Comment by Friptick** [comment]
OP is right that the Podboat download-queue workflow seems fastidious when `wget` exists. More resilient, yes, but also more cumbersome than just "download it".

**But there's another problem!** What if the user wants to rewrite the downloaded file's metadata using the feed values? In other words, to correctly populate the ID3 tags on a downloaded MP3 so that it is decipherable in an audio player. Many podcast RSS feeds contain poorly tagged enclosures, presumably because it's assumed they'll only ever be played by a podcast client within the context of their RSS feed. **The Podboat workflow means that any such missing data is lost irretrievably.**

The obvious solution would be a function to pipe out the enclosure URL, together with the metadata which might not be in the enclosure - "Feed name", "Article title", etc. Either just to `wget` or to a script which can make use of the RSS metadata to rewrite tags and so on.

Is there any hope that such a pipe function might be implemented?

**Comment by Minoru** [comment]
It's open source, so there's always hope :) In #1691, I outline how enqueueing might be unified with some other features, but no work has been done on that yet. Patches are always welcome (though that one requires a design discussion first, I guess).

**Comment by Friptick** [comment]
> It's open source, so there's always hope

Indeed. I was being disingenuous. I actually have much expertise in the art of nagging and pestering software maintainers. :)

Your little manifesto in #1691 is persuasive, nice work for thinking it through.

I am new to this project but it does feel like a relative of Mutt. And Mutt allows direct download of email attachments.
